fem/libceed/convection.h

```diff
@@ -0,0 +1,523 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+/// A structure used to pass additional data to f_build_conv and f_apply_conv
+struct BuildContext { CeedInt dim, space_dim, vdim; CeedScalar coeff[3]; };
+
+/// libCEED Q-function for building quadrature data for a convection operator
+/// with a constant coefficient
+CEED_QFUNCTION(f_build_conv_const)(void *ctx, CeedInt Q,
+                                   const CeedScalar *const *in,
+                                   CeedScalar *const *out)
+{
+   BuildContext *bc = (BuildContext*)ctx;
+   // in[0] is Jacobians with shape [dim, nc=dim, Q]
+   // in[1] is quadrature weights, size (Q)
+   //
+   // At every quadrature point, compute qw/det(J).adj(J).adj(J)^T and store
+   // the symmetric part of the result.
+   const CeedScalar coeff0 = bc->coeff[0];
+   const CeedScalar coeff1 = bc->coeff[1];
+   const CeedScalar coeff2 = bc->coeff[2];
+   const CeedScalar *J = in[0], *qw = in[1];
+   CeedScalar *qd = out[0];
+   switch (bc->dim + 10 * bc->space_dim)
+   {
+      case 11:
+         for (CeedInt i = 0; i < Q; i++)
+         {
+            qd[i] = coeff0 * qw[i] * J[i];
+         }
+         break;
+      case 22:
+         for (CeedInt i = 0; i < Q; i++)
+         {
+            // J: 0 2   qd: 0 1   adj(J):  J22 -J12
+            //    1 3       1 2           -J21  J11
+            const CeedScalar J11 = J[i + Q * 0];
+            const CeedScalar J21 = J[i + Q * 1];
+            const CeedScalar J12 = J[i + Q * 2];
+            const CeedScalar J22 = J[i + Q * 3];
+            const CeedScalar w = - qw[i]; // TODO remove - and use alpha
```

Comment:
Will's suggestion is better, but if you don't do that I recommend putting `w = alpha * wq[i];` here (and in all similar lines) and `constexpr double alpha = -1.0` at the top of the function or something similar. That way future extension / maintainability is easier even if we don't fix the alpha issue in this PR.